docs/proposals/06-immutable-etcd-backups.md

```diff
@@ -0,0 +1,315 @@
+
+### Goals
+
+- Implement immutable backup support for ETCD clusters.
+- Secure backup data against unintended or unauthorized modifications after creation.
+- Implement changes required in `etcd-backup-restore` and `etcd-druid` to support this proposal.
+
+### Non-Goals
+
+- Implementing object-level immutability policies at this stage.
+- Supporting immutable backups on storage providers that do not offer immutability features (e.g., OpenStack Swift).
+
+## Proposal
+
+### Overview
+
+We propose introducing immutability in backup storage by leveraging cloud provider features that support a write-once-read-many (WORM) model. This approach will prevent data alterations post-creation, enhancing data integrity and security.
+
+There are two types of immutability options to consider:
+
+1. **Bucket-Level Immutability:** Applies a uniform immutability period to all objects within a bucket. This is widely supported and easier to manage across different cloud providers.
+2. **Object-Level Immutability:** Allows setting immutability periods on a per-object basis, offering more granular control but with increased complexity and varying support across providers.
+
+In the detailed design, we will focus on bucket-level immutability policies due to their broader support and simpler management.
+
+### Detailed Design
+
+#### Bucket Immutability Mechanism
+
+The bucket immutability feature configures an immutability policy for a cloud storage bucket, dictating how long objects in the bucket must remain immutable. It also allows for locking the bucket's immutability policy, permanently preventing the policy from being reduced or removed.
+
+- **Supported by Major Providers:**
+  - **Google Cloud Storage (GCS):** [Bucket Lock](https://cloud.google.com/storage/docs/bucket-lock)
+  - **Amazon S3 (S3):** [Object Lock](https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-lock.html)
+  - **Azure Blob Storage (ABS):** [Immutable Blob Storage](https://learn.microsoft.com/en-us/azure/storage/blobs/immutable-policy-configure-container-scope?tabs=azure-portal)
+
+- **Not Supported:**
+  - **OpenStack Swift**
+
+**Implementation Details:**
+
+- Operators are responsible for configuring new or existing buckets with these immutability settings before `etcd-backup-restore` begins uploading snapshots.
+- Once configured, snapshots uploaded by `etcd-backup-restore` will also be immutable and cannot be altered or deleted until the immutability period expires.
+- No additional configuration needs to be passed to etcd-druid.
+
+#### ETCD Backup Configuration
+
+Operators must ensure that the ETCD backup configuration aligns with the immutability requirements, including setting appropriate immutability periods.
```

Comment:
Yes, we have referred to the links for configuring immutable backup buckets for both standalone and Gardener cases. We have removed the compaction approach since both the compaction and re-uploading approaches use the operator framework and take full snapshots in the same way. The only difference is starting the embedded `etcd`. The more appropriate approach is re-uploading. Therefore, we have removed the compaction approach as it is redundant, doing the same thing apart from starting the embedded `etcd` and compacting.